rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/ParametricTextDocumentService.java

```diff
@@ -597,38 +605,53 @@ private static <T> T last(List<T> l) {
         return l.get(l.size() - 1);
     }
 
+    private boolean isLanguageRegistered(ISourceLocation loc) {
+        return registeredExtensions.containsKey(extension(loc));
+    }
+
+    private String language(ISourceLocation loc) {
+        var ext = extension(loc);
+        var language = registeredExtensions.get(ext);
+        if (language == null) {
+            throw new UnsupportedOperationException(String.format("Rascal Parametric LSP has no support for this file, since no language is registered for extension '%s': %s", ext, loc));
+        }
+        return language;
+    }
+
     private ILanguageContributions contributions(ISourceLocation doc) {
-        String language = registeredExtensions.get(extension(doc));
-        if (language != null) {
-            ILanguageContributions contrib = contributions.get(language);
+        ILanguageContributions contrib = contributions.get(language(doc));
 
-            if (contrib != null) {
-                return contrib;
-            }
+        if (contrib == null) {
+            throw new UnsupportedOperationException("Rascal Parametric LSP has no support for this file: " + doc);
         }
 
-        throw new UnsupportedOperationException("Rascal Parametric LSP has no support for this file: " + doc);
+        return contrib;
     }
 
     private static String extension(ISourceLocation doc) {
         return URIUtil.getExtension(doc);
     }
 
     private ParametricFileFacts facts(ISourceLocation doc) {
-        String language = registeredExtensions.get(extension(doc));
-        if (language != null) {
-            ParametricFileFacts fact = facts.get(language);
-            if (fact != null) {
-                return fact;
-            }
+        ParametricFileFacts fact = facts.get(language(doc));
+
+        if (fact == null) {
+            throw new UnsupportedOperationException("Rascal Parametric LSP has no support for this file: " + doc);
         }
-        throw new UnsupportedOperationException("Rascal Parametric LSP has no support for this file: " + doc);
+
+        return fact;
     }
 
     private TextDocumentState open(TextDocumentItem doc, long timestamp) {
-        return files.computeIfAbsent(Locations.toLoc(doc),
-            l -> new TextDocumentState(contributions(l)::parsing, l, doc.getVersion(), doc.getText(), timestamp)
-        );
+        var loc = Locations.toLoc(doc);
+        BiFunction<ISourceLocation, String, CompletableFuture<ITree>> parser
+            = isLanguageRegistered(loc)
+            ? contributions(loc)::parsing
+            // Language not (yet) registered; plug in an incomplete parse future
+            : (l, s) -> new CompletableFuture<>();
```

Comment:
this future will never complete, won't that cause issues? like any down stream jobs?